## Noom Data Anlaysis


<img src="https://www.dropbox.com/s/b0gsp4jncwwnoq0/noom-coach-logo.jpg?dl=1" alt="drawing" width="200px"/>


In [6]:
import pandas as pd

In [7]:

raw_data = pd.read_csv("noom/noom_user.csv", parse_dates=["Purchased At"])


print(raw_data.columns)
print(raw_data.shape)
raw_data.head()

Index(['Access Code', 'Name', 'Gender', 'Age', 'Height', 'Initial Weight',
       'Lowest Weight', 'Target Weight', 'Product Name', 'Status', 'Price',
       'Purchased At', 'Payment Type', 'Channel'],
      dtype='object')
(10000, 14)


Access Code Name  Gender   Age  Height  Initial Weight  Lowest Weight  \
0    Y9RY2VSI  김승혜  FEMALE  25.0   172.0            66.9           65.8   
1    3GTN3S3B  허승준    MALE  26.0   176.0            70.0            NaN   
2    6B0IG276  이지민  FEMALE  23.0   171.0            98.0            NaN   
3    EMGRU2MO  장설윤  FEMALE  20.0   160.0            70.7            NaN   
4    1ELG96TX  서성빈  FEMALE  28.0   165.0            55.5            NaN   

   Target Weight         Product Name     Status   Price  \
0      55.000000          눔 체중감량 프로그램  completed  112500   
1      65.000000          눔 체중감량 프로그램  completed   44780   
2      91.140000  눔 체중감량 프로그램 (천원 체험)  completed  132000   
3      53.000000  눔 체중감량 프로그램 (천원 체험)  completed  112500   
4      51.615002          눔 체중감량 프로그램  completed   44780   

             Purchased At Payment Type   Channel  
0 2017-04-14 19:03:29.976    Recurring    others  
1 2017-05-23 20:53:54.368    Recurring    others  
2 2017-08-23 23:39:21.840    Recurring  facebook  
3 2017-08-28 20:18:22.824    Recurring     naver  
4 2017-05-07 17:50:30.944    Recurring  facebook

<table border="1" class="dataframe">
    <thead>
        <tr style="text-align: right;">
            <th></th>
            <th>Name</th>
            <th>Gender</th>
            <th>Age</th>
            <th>Height</th>
            <th>Initial Weight</th>
            <th>Lowest Weight</th>
            <th>Target Weight</th>
            <th>Status</th>
            <th>Price</th>
            <th>Purchased At</th>
            <th>Channel</th>
        </tr>
        <tr>
            <th>Access Code</th>
            <th></th>
            <th></th>
            <th></th>
            <th></th>
            <th></th>
            <th></th>
            <th></th>
            <th></th>
            <th></th>
            <th></th>
            <th></th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <th>Y9RY2VSI</th>
            <td>김승혜</td>
            <td>FEMALE</td>
            <td>25.0</td>
            <td>172.0</td>
            <td>66.9</td>
            <td>65.8</td>
            <td>55.000000</td>
            <td>completed</td>
            <td>112500</td>
            <td>2017-04-14 19:03:29.976</td>
            <td>others</td>
        </tr>
        <tr>
            <th>3GTN3S3B</th>
            <td>허승준</td>
            <td>MALE</td>
            <td>26.0</td>
            <td>176.0</td>
            <td>70.0</td>
            <td>NaN</td>
            <td>65.000000</td>
            <td>completed</td>
            <td>44780</td>
            <td>2017-05-23 20:53:54.368</td>
            <td>others</td>
        </tr>
        <tr>
            <th>6B0IG276</th>
            <td>이지민</td>
            <td>FEMALE</td>
            <td>23.0</td>
            <td>171.0</td>
            <td>98.0</td>
            <td>NaN</td>
            <td>91.140000</td>
            <td>completed</td>
            <td>132000</td>
            <td>2017-08-23 23:39:21.840</td>
            <td>facebook</td>
        </tr>
        <tr>
            <th>EMGRU2MO</th>
            <td>장설윤</td>
            <td>FEMALE</td>
            <td>20.0</td>
            <td>160.0</td>
            <td>70.7</td>
            <td>NaN</td>
            <td>53.000000</td>
            <td>completed</td>
            <td>112500</td>
            <td>2017-08-28 20:18:22.824</td>
            <td>naver</td>
        </tr>
        <tr>
            <th>1ELG96TX</th>
            <td>서성빈</td>
            <td>FEMALE</td>
            <td>28.0</td>
            <td>165.0</td>
            <td>55.5</td>
            <td>NaN</td>
            <td>51.615002</td>
            <td>completed</td>
            <td>44780</td>
            <td>2017-05-07 17:50:30.944</td>
            <td>facebook</td>
        </tr>
    </tbody>
</table>

In [3]:
# Write your code here!

columns  = [
    "Access Code",
"Name",
"Gender",
"Age",
"Height",
"Initial Weight",
"Lowest Weight",
"Target Weight",
"Status",
"Price",
"Purchased At",
"Channel"
]
#df = raw_data[columns].copy()
df = raw_data[columns]
df = df.set_index("Access Code")
df.head()

Name  Gender   Age  Height  Initial Weight  Lowest Weight  \
Access Code                                                             
Y9RY2VSI     김승혜  FEMALE  25.0   172.0            66.9           65.8   
3GTN3S3B     허승준    MALE  26.0   176.0            70.0            NaN   
6B0IG276     이지민  FEMALE  23.0   171.0            98.0            NaN   
EMGRU2MO     장설윤  FEMALE  20.0   160.0            70.7            NaN   
1ELG96TX     서성빈  FEMALE  28.0   165.0            55.5            NaN   

             Target Weight     Status   Price            Purchased At  \
Access Code                                                             
Y9RY2VSI         55.000000  completed  112500 2017-04-14 19:03:29.976   
3GTN3S3B         65.000000  completed   44780 2017-05-23 20:53:54.368   
6B0IG276         91.140000  completed  132000 2017-08-23 23:39:21.840   
EMGRU2MO         53.000000  completed  112500 2017-08-28 20:18:22.824   
1ELG96TX         51.615002  completed   44780 2017-05-07 17:50:30.944   

              Channel  
Access Code            
Y9RY2VSI       others  
3GTN3S3B       others  
6B0IG276     facebook  
EMGRU2MO        naver  
1ELG96TX     facebook

In [4]:
raw_data["Gender"].unique()

array(['FEMALE', 'MALE', nan], dtype=object)

In [5]:
# Write your code here!
df['Gender(clean)']= df['Gender'].str.lower()
df['Gender(clean)'].head()

Access Code
Y9RY2VSI    female
3GTN3S3B      male
6B0IG276    female
EMGRU2MO    female
1ELG96TX    female
Name: Gender(clean), dtype: object

In [6]:
df[['Gender', 'Gender(clean)']].head()
df['Gender(clean)'].value_counts()

female    8846
male      1023
Name: Gender(clean), dtype: int64

In [7]:
# Height 컬럼에서 가장 작은(min) 값을 찾아와서 출력합니다.
raw_data["Height"].min().astype


<function float64.astype>

In [8]:
# Write your code here!
import numpy as np
#df['Height(clean)'] = df['Height'].replace(-1, np.nan)
df['Height(clean)'] = df['Height']
df.loc[df['Height'] == -1, 'Height(clean)'] = np.nan
df.loc[df['Height'] == -1, ['Height', 'Height(clean)']].head()

Height  Height(clean)
Access Code                       
O4OWMJG7       -1.0            NaN
H6EV5AXL       -1.0            NaN
O1IAZS7A       -1.0            NaN
5NEQOWHW       -1.0            NaN
OFAXUNXD       -1.0            NaN

In [9]:
# Print( min, max, mean by Gender)
print("Min: ",df['Height(clean)'].min())
print("Max: ",df['Height(clean)'].max())
# df.groupby("Gender(clean)")['Height(clean)'].mean()
pd.pivot_table(df, index="Gender(clean)", values = 'Height(clean)')


Min:  106.0
Max:  203.2


Height(clean)
Gender(clean)               
female            162.116913
male              175.831965

In [10]:
raw_data["Age"].min()

0.0

In [11]:
raw_data["Age"].max()

173.0

In [12]:
'''나이가 0인 데이터는 NaN으로 변경
나이가 60세 이상인 데이터는 NaN으로 변경'''
df['Age(clean)'] = df['Age']
df.loc[df['Age'] > 59, 'Age(clean)'] = np.nan
df['Age(clean)'] = df['Age(clean)'].replace(0, np.nan)


In [13]:
# Checking 
df.loc[(df['Age'] > 59) | (df['Age'] == 0) , ['Age', 'Age(clean)']].head()


Age  Age(clean)
Access Code                  
9PTGVW4B      0.0         NaN
ACV6D35S      0.0         NaN
Y0OK1FWA     85.0         NaN
WX34HGBL      0.0         NaN
R9XCS81F      0.0         NaN

In [14]:
''' 1) 최소/최대/평균 나이를 구하고, 2) 남성/여성별 평균 나이를 구해주세요.
가령 현재 나이가 가장 어린 사용자는 13세이며, 반대로 가장 나이가 많은 사용자는 59세, 
그리고 평균 나이는 약 27.38세 입니다. 남성의 평균 나이와 여성의 평균 나이는 다음과 같습니다.'''
print("max:", df['Age(clean)'].max(), 
      "min:", df['Age(clean)'].min(),
      "mean:", df['Age(clean)'].mean())

# df.groupby("Gender(clean)")['Age(clean)'].mean()
pd.pivot_table(df, index="Gender(clean)", values = 'Age(clean)')


max: 59.0 min: 13.0 mean: 27.39381024860477


Age(clean)
Gender(clean)            
female          27.172929
male            29.309127

### Define VIP Customer  

In [15]:
col = ["Name",
"Age(clean)",
"Height(clean)",
"Initial Weight",
"Lowest Weight",
"Target Weight",
"Status"]
df_copy = df[col].copy()
df_copy.head()

Name  Age(clean)  Height(clean)  Initial Weight  Lowest Weight  \
Access Code                                                                  
Y9RY2VSI     김승혜        25.0          172.0            66.9           65.8   
3GTN3S3B     허승준        26.0          176.0            70.0            NaN   
6B0IG276     이지민        23.0          171.0            98.0            NaN   
EMGRU2MO     장설윤        20.0          160.0            70.7            NaN   
1ELG96TX     서성빈        28.0          165.0            55.5            NaN   

             Target Weight     Status  
Access Code                            
Y9RY2VSI         55.000000  completed  
3GTN3S3B         65.000000  completed  
6B0IG276         91.140000  completed  
EMGRU2MO         53.000000  completed  
1ELG96TX         51.615002  completed


$$
BMI = \frac{체중(kg)}{키(m) \times 키(m)}
$$



In [16]:
# Write your code here!
#1. Weight Loss(goal)
df_copy['Weight Loss(goal)'] = df['Initial Weight'] - df['Target Weight']


In [17]:
#2. Weight Loss(current)
df_copy['Weight Loss(current)'] = df['Initial Weight'] - df['Lowest Weight']


In [18]:
#3. 체질량지수(BMI)
height_meter = df_copy['Height(clean)'] / 100
df_copy['BMI'] = df_copy['Initial Weight'] /(height_meter * height_meter)

In [19]:
# checking
df_copy[["Name","Weight Loss(goal)", "Weight Loss(current)", "BMI"]].head()

Name  Weight Loss(goal)  Weight Loss(current)        BMI
Access Code                                                         
Y9RY2VSI     김승혜          11.900000                   1.1  22.613575
3GTN3S3B     허승준           5.000000                   NaN  22.598140
6B0IG276     이지민           6.860000                   NaN  33.514586
EMGRU2MO     장설윤          17.700000                   NaN  27.617187
1ELG96TX     서성빈           3.884998                   NaN  20.385675

In [20]:
# Write your code here!
def invalid_finder(row):
    if (pd.isnull(row['Age(clean)'])) | (pd.isnull(row["Height(clean)"])) | (pd.isnull(row["Initial Weight"])) | (pd.isnull(row["Target Weight"])) | (pd.isnull(row["Lowest Weight"])):
        return True
    elif (row["Height(clean)"] > 200) or (row["Height(clean)"] < 140):
        return True
    elif (row['BMI'] <18.5) |( row['BMI'] >30.0):
        return True
    elif row['Weight Loss(goal)'] <0:
        return True
    else:
        return False

In [21]:
df_copy['Invalid']=df_copy.apply(invalid_finder, axis ='columns')

In [22]:
# 눔의 프로그램을 결제한 구매자. (Status == "completed") False
df_copy.loc[~(df['Status'] == "completed"), "Invalid"] = False

In [23]:
# checking
df_copy[df_copy['Invalid'] == True].head()

Name  Age(clean)  Height(clean)  Initial Weight  Lowest Weight  \
Access Code                                                                  
3GTN3S3B     허승준        26.0          176.0            70.0            NaN   
6B0IG276     이지민        23.0          171.0            98.0            NaN   
EMGRU2MO     장설윤        20.0          160.0            70.7            NaN   
1ELG96TX     서성빈        28.0          165.0            55.5            NaN   
U9AELHAT     배소정        22.0          161.0            76.0            NaN   

             Target Weight     Status  Weight Loss(goal)  \
Access Code                                                
3GTN3S3B         65.000000  completed           5.000000   
6B0IG276         91.140000  completed           6.860000   
EMGRU2MO         53.000000  completed          17.700000   
1ELG96TX         51.615002  completed           3.884998   
U9AELHAT         48.000000  completed          28.000000   

             Weight Loss(current)        BMI  Invalid  
Access Code                                            
3GTN3S3B                      NaN  22.598140     True  
6B0IG276                      NaN  33.514586     True  
EMGRU2MO                      NaN  27.617187     True  
1ELG96TX                      NaN  20.385675     True  
U9AELHAT                      NaN  29.319856     True

In [24]:
# check if total number of data = 2619 
df_copy[df_copy['Invalid'] == True].shape

(2619, 11)

In [26]:
# Write your code here!
df_copy['VIP'] = False

In [27]:
# all conitions must be satisfied
def VIP_finder(row):
    if (~pd.isnull(row['Weight Loss(goal)'])) & (~pd.isnull(row["Weight Loss(current)"])) & (~pd.isnull(row["BMI"])) & (row["Weight Loss(current)"] > 10) & (row['BMI'] >30.0) & (row['Weight Loss(current)'] > row['Weight Loss(goal)']) :
        return True
    else:
        return False

In [28]:
df_copy['VIP'] = df_copy.apply(VIP_finder, axis ='columns')
df_copy.loc[~(df_copy['Status'] == "completed"),'VIP'] = False

In [29]:
# checking
df_copy[df_copy['VIP'] == True]

Name  Age(clean)  Height(clean)  Initial Weight  Lowest Weight  \
Access Code                                                                  
3T1I8I8E     임솔지        23.0          158.0         80.0137           54.5   
PJYKU9OW     홍윤오        31.0          174.0         99.9000           76.6   
0EMTSGLJ     류선정        34.0          167.0         86.0000           73.2   
FBEAIFW0     서서원        23.0          170.0         95.0000           75.7   
8QQV2YDW     홍서율        23.0          170.0         95.0000           75.7   
99KOLRU8     고서연        26.0          162.0        106.0000           95.2   
IBOWZ9WZ     손서애        22.0          166.0         83.0000           66.1   
6EH2LGR5     문세영        34.0          164.0         88.8000           56.3   
QQLYGTWD     황수윤        26.0          165.0        105.0000           87.8   
4Z1WB3UZ     허지예        26.0          162.0        106.0000           95.2   
2YAKET8R     서승희        30.0          164.0         99.8000           86.4   
LDPPDM0M     윤지안        19.0          168.0         93.1000           65.5   
PKHJWII8     정선영        23.0          170.0         95.0000           75.7   
3B3WQA4A     홍슬비        28.0          158.0         84.0000           69.3   
SDY4VS0P     오채현         NaN          170.0        103.1000           57.1   

             Target Weight     Status  Weight Loss(goal)  \
Access Code                                                
3T1I8I8E         77.745742  completed           2.267958   
PJYKU9OW         84.000000  completed          15.900000   
0EMTSGLJ         80.050003  completed           5.949997   
FBEAIFW0         85.000000  completed          10.000000   
8QQV2YDW         85.000000  completed          10.000000   
99KOLRU8         98.580000  completed           7.420000   
IBOWZ9WZ         73.000000  completed          10.000000   
6EH2LGR5         82.584003  completed           6.215997   
QQLYGTWD         97.650000  completed           7.350000   
4Z1WB3UZ         98.580000  completed           7.420000   
2YAKET8R         92.814000  completed           6.986000   
LDPPDM0M         69.000000  completed          24.100000   
PKHJWII8         85.000000  completed          10.000000   
3B3WQA4A         75.000000  completed           9.000000   
SDY4VS0P        100.099998  completed           3.000001   

             Weight Loss(current)        BMI  Invalid   VIP  
Access Code                                                  
3T1I8I8E                  25.5137  32.051634     True  True  
PJYKU9OW                  23.3000  32.996433     True  True  
0EMTSGLJ                  12.8000  30.836531     True  True  
FBEAIFW0                  19.3000  32.871972     True  True  
8QQV2YDW                  19.3000  32.871972     True  True  
99KOLRU8                  10.8000  40.390184     True  True  
IBOWZ9WZ                  16.9000  30.120482     True  True  
6EH2LGR5                  32.5000  33.016062     True  True  
QQLYGTWD                  17.2000  38.567493     True  True  
4Z1WB3UZ                  10.8000  40.390184     True  True  
2YAKET8R                  13.4000  37.105889     True  True  
LDPPDM0M                  27.6000  32.986111     True  True  
PKHJWII8                  19.3000  32.871972     True  True  
3B3WQA4A                  14.7000  33.648454     True  True  
SDY4VS0P                  46.0000  35.674740     True  True

In [30]:
df_copy[df_copy['VIP'] == True].shape

(15, 12)

In [31]:
# .value_counts() 를 통해 Status의 종류와 그 누적 갯수를 출력합니다.
df_copy["Status"].value_counts()

completed    5400
cancelled    4010
refunded      590
Name: Status, dtype: int64

In [32]:
# Write your code here!
df['Age(Group)'] = np.nan

In [33]:
def age_group(row):
    if (row['Age(clean)'] >= 55) & (row['Age(clean)'] <= 99):
        return "55~99"
    elif (row['Age(clean)'] >= 18) & (row['Age(clean)'] <= 24):
        return "18~24"
    elif (row['Age(clean)'] >= 25) & (row['Age(clean)'] <= 35):
        return "25~35"
    elif (row['Age(clean)'] >= 36) & (row['Age(clean)'] <= 44):
        return "36~44"
    elif (row['Age(clean)'] >= 45) & (row['Age(clean)'] <= 54):
        return "45~54"
    elif (row['Age(clean)'] >= 0) & (row['Age(clean)'] <= 17):
        return "00~17"

In [34]:
df['count'] = int(1)

In [35]:
# Age(Group) needed
df['Age(Group)']=df.apply(age_group, axis=1) # axis = "columns"

In [36]:
df_rate =pd.pivot_table(df, index=['Gender(clean)', 'Age(Group)'], columns = "Status", values='count',aggfunc = np.sum)

In [37]:
def null_changer(row):
    row.loc[row['cancelled'].isnull(),'cancelled'] = 0
    row.loc[row['completed'].isnull(),'completed'] = 0
    row.loc[row['refunded'].isnull(),'refunded'] = 0

In [38]:
# null to int(0)
null_changer(df_rate)

In [39]:
# creating total and coversion
df_rate['total'] = df_rate['cancelled'] + df_rate["completed"] + df_rate["refunded"]
df_rate['conversion'] = df_rate['completed'] / df_rate['total']

In [40]:
# checking 
df_rate

Status                    cancelled  completed  refunded   total  conversion
Gender(clean) Age(Group)                                                    
female        00~17            25.0       35.0       3.0    63.0    0.555556
              18~24          1637.0     1827.0     149.0  3613.0    0.505674
              25~35          1664.0     2288.0     271.0  4223.0    0.541795
              36~44           206.0      421.0      46.0   673.0    0.625557
              45~54            74.0      160.0      25.0   259.0    0.617761
              55~99             0.0        5.0       0.0     5.0    1.000000
male          00~17             3.0        1.0       0.0     4.0    0.250000
              18~24            80.0      100.0      11.0   191.0    0.523560
              25~35           235.0      404.0      57.0   696.0    0.580460
              36~44            21.0       72.0       9.0   102.0    0.705882
              45~54             9.0       13.0       3.0    25.0    0.520000
              55~99             0.0        1.0       0.0     1.0    1.000000

In [41]:
df["Purchased At"] = pd.to_datetime(df["Purchased At"])

In [42]:
df["Purchased At(hour)"] = df["Purchased At"].dt.hour

In [43]:
# 0시 ~ 23시 사이의 결제/캔슬/환불 비율.
df_rate_hour =pd.pivot_table(df, index=['Purchased At(hour)'], columns = "Status", values='count',aggfunc = np.sum)
# creating total and coversion
df_rate_hour['total'] = df_rate_hour['cancelled'] + df_rate_hour["completed"] + df_rate_hour["refunded"]
df_rate_hour['conversion'] = df_rate_hour['completed'] / df_rate_hour['total']

In [44]:
# checking 
df_rate_hour.head()

Status              cancelled  completed  refunded  total  conversion
Purchased At(hour)                                                   
0                         236        344        26    606    0.567657
1                         156        207        28    391    0.529412
2                          90         97         4    191    0.507853
3                          58         66         5    129    0.511628
4                          59         45         7    111    0.405405

In [45]:
df["Purchased At(weekday)"] = df["Purchased At"].dt.weekday

In [46]:
# Write your code here!
# 월요일-일요일 사이의 결제/캔슬/환불 비율.
df_rate_weekday =pd.pivot_table(df, index=['Purchased At(weekday)'], columns = "Status", values='count',aggfunc = np.sum)
# creating total and coversion
df_rate_weekday['total'] = df_rate_weekday['cancelled'] + df_rate_weekday["completed"] + df_rate_weekday["refunded"]
df_rate_weekday['conversion'] = df_rate_weekday['completed'] / df_rate_weekday['total']

In [47]:
df_rate_weekday

Status                 cancelled  completed  refunded  total  conversion
Purchased At(weekday)                                                   
0                            691        863        93   1647    0.523983
1                            694        935       102   1731    0.540150
2                            679        953        90   1722    0.553426
3                            616        813        88   1517    0.535926
4                            490        674        56   1220    0.552459
5                            412        537        73   1022    0.525440
6                            428        625        88   1141    0.547765

In [48]:
index = ['Monday', 'Tuesday','Wedneday','Thursday','Friday', 'Saturday','Sunday']
df_rate_weekday.index = index
# checking
df_rate_weekday

Status    cancelled  completed  refunded  total  conversion
Monday          691        863        93   1647    0.523983
Tuesday         694        935       102   1731    0.540150
Wedneday        679        953        90   1722    0.553426
Thursday        616        813        88   1517    0.535926
Friday          490        674        56   1220    0.552459
Saturday        412        537        73   1022    0.525440
Sunday          428        625        88   1141    0.547765

In [49]:
df["Channel"].value_counts()

facebook     6880
others       1390
naver        1009
direct        297
email         271
google        120
instagram      33
Name: Channel, dtype: int64

In [50]:
# Write your code here!
df_channel =pd.pivot_table(df, index=['Channel'], columns = "Status", values='count',aggfunc = np.sum)
# creating total and coversion
df_channel['total'] = df_channel['cancelled'] + df_channel["completed"] + df_channel["refunded"]
df_channel['conversion'] = df_channel['completed'] / df_channel['total']

In [51]:
# checking 
df_channel

Status     cancelled  completed  refunded  total  conversion
Channel                                                     
direct           119        169         9    297    0.569024
email             93        155        23    271    0.571956
facebook        2812       3654       414   6880    0.531105
google            42         66        12    120    0.550000
instagram         13         17         3     33    0.515152
naver            386        568        55   1009    0.562934
others           545        771        74   1390    0.554676

In [52]:
# 먼저 코칭 데이터를 가져옵니다.
# 이 결과를 coach라는 이름의 변수에 할당합니다.
coach = pd.read_csv("noom/noom_coach.csv", index_col="Access Code")


print(coach.shape)

# coach 데이터의 상위 5개를 출력합니다.
coach.head()

(10000, 100)


정은오 코치(VEV4PGJB)  오승혁 코치(VENPKBP9)  조소은 코치(D0WASBXN)  \
Access Code                                                         
Y9RY2VSI                    0                 0                 0   
3GTN3S3B                    0                 0                 0   
6B0IG276                    0                 0                 0   
EMGRU2MO                    0                 0                 0   
1ELG96TX                    0                 0                 0   

             고영재 코치(C91AVNGB)  조수민 코치(OBCAO3W0)  강채아 코치(WH2NIKCO)  \
Access Code                                                         
Y9RY2VSI                    0                 0                 0   
3GTN3S3B                    0                 0                 0   
6B0IG276                    0                 0                 0   
EMGRU2MO                    0                 0                 0   
1ELG96TX                    0                 0                 0   

             황다훈 코치(1I6IWURH)  백슬은 코치(228BFB50)  유채우 코치(IW53Y9AW)  \
Access Code                                                         
Y9RY2VSI                    0                 0                 0   
3GTN3S3B                    0                 0                 0   
6B0IG276                    0                 0                 0   
EMGRU2MO                    1                 0                 0   
1ELG96TX                    0                 0                 0   

             송지선 코치(WL0877P7)  ...  오초빈 코치(A3WOLAQM)  서수정 코치(F36LORFC)  \
Access Code                    ...                                       
Y9RY2VSI                    0  ...                 0                 0   
3GTN3S3B                    0  ...                 0                 0   
6B0IG276                    0  ...                 0                 0   
EMGRU2MO                    0  ...                 0                 0   
1ELG96TX                    0  ...                 0                 0   

             정서율 코치(LX1G7EMD)  고우재 코치(SKNL9Z4P)  문한규 코치(OU1WVDGA)  \
Access Code                                                         
Y9RY2VSI                    0                 1                 0   
3GTN3S3B                    0                 0                 1   
6B0IG276                    0                 0                 0   
EMGRU2MO                    0                 0                 0   
1ELG96TX                    0                 0                 0   

             황세안 코치(3QUBQAVE)  홍성은 코치(2I3QJQ5O)  고성은 코치(34T7XPYR)  \
Access Code                                                         
Y9RY2VSI                    0                 0                 0   
3GTN3S3B                    0                 0                 0   
6B0IG276                    0                 0                 0   
EMGRU2MO                    0                 0                 0   
1ELG96TX                    0                 0                 0   

             백한율 코치(HPWAN8R0)  안슬은 코치(QAVWJSZ1)  
Access Code                                      
Y9RY2VSI                    0                 0  
3GTN3S3B                    0                 0  
6B0IG276                    0                 0  
EMGRU2MO                    0                 0  
1ELG96TX                    0                 0  

[5 rows x 100 columns]

In [53]:
# Write your code here!
# cheching
df_concat_coach =pd.concat([df[['Name', 'Status']], coach], axis= 1)
df_concat_coach.head()

Name     Status  정은오 코치(VEV4PGJB)  오승혁 코치(VENPKBP9)  \
Access Code                                                       
Y9RY2VSI     김승혜  completed                 0                 0   
3GTN3S3B     허승준  completed                 0                 0   
6B0IG276     이지민  completed                 0                 0   
EMGRU2MO     장설윤  completed                 0                 0   
1ELG96TX     서성빈  completed                 0                 0   

             조소은 코치(D0WASBXN)  고영재 코치(C91AVNGB)  조수민 코치(OBCAO3W0)  \
Access Code                                                         
Y9RY2VSI                    0                 0                 0   
3GTN3S3B                    0                 0                 0   
6B0IG276                    0                 0                 0   
EMGRU2MO                    0                 0                 0   
1ELG96TX                    0                 0                 0   

             강채아 코치(WH2NIKCO)  황다훈 코치(1I6IWURH)  백슬은 코치(228BFB50)  ...  \
Access Code                                                        ...   
Y9RY2VSI                    0                 0                 0  ...   
3GTN3S3B                    0                 0                 0  ...   
6B0IG276                    0                 0                 0  ...   
EMGRU2MO                    0                 1                 0  ...   
1ELG96TX                    0                 0                 0  ...   

             오초빈 코치(A3WOLAQM)  서수정 코치(F36LORFC)  정서율 코치(LX1G7EMD)  \
Access Code                                                         
Y9RY2VSI                    0                 0                 0   
3GTN3S3B                    0                 0                 0   
6B0IG276                    0                 0                 0   
EMGRU2MO                    0                 0                 0   
1ELG96TX                    0                 0                 0   

             고우재 코치(SKNL9Z4P)  문한규 코치(OU1WVDGA)  황세안 코치(3QUBQAVE)  \
Access Code                                                         
Y9RY2VSI                    1                 0                 0   
3GTN3S3B                    0                 1                 0   
6B0IG276                    0                 0                 0   
EMGRU2MO                    0                 0                 0   
1ELG96TX                    0                 0                 0   

             홍성은 코치(2I3QJQ5O)  고성은 코치(34T7XPYR)  백한율 코치(HPWAN8R0)  \
Access Code                                                         
Y9RY2VSI                    0                 0                 0   
3GTN3S3B                    0                 0                 0   
6B0IG276                    0                 0                 0   
EMGRU2MO                    0                 0                 0   
1ELG96TX                    0                 0                 0   

             안슬은 코치(QAVWJSZ1)  
Access Code                    
Y9RY2VSI                    0  
3GTN3S3B                    0  
6B0IG276                    0  
EMGRU2MO                    0  
1ELG96TX                    0  

[5 rows x 102 columns]

In [54]:
# Write your code here!
# see data first => 0 or 1 
df_concat_coach["정은오 코치(VEV4PGJB)"].unique()

array([0, 1])

In [55]:
df_coach = pd.pivot_table(df_concat_coach, columns = "Status",aggfunc = np.sum)

In [56]:
index = ["정은오 코치(VEV4PGJB)",
       "오승혁 코치(VENPKBP9)","조소은 코치(D0WASBXN)",
       "고영재 코치(C91AVNGB)","조수민 코치(OBCAO3W0)"]

In [57]:
# checking 
df_coach.loc[index]

Status            cancelled  completed  refunded
정은오 코치(VEV4PGJB)         18         23         3
오승혁 코치(VENPKBP9)         14         44         9
조소은 코치(D0WASBXN)         12         12         1
고영재 코치(C91AVNGB)         22         27         4
조수민 코치(OBCAO3W0)         49         46         8

In [61]:
# Write your code here!
# creating total and coversion
df_coach['total'] = df_coach['cancelled'] + df_coach["completed"] + df_coach["refunded"]
df_coach['conversion rate'] = df_coach['completed'] / df_coach['total']
df_coach['concellation rate'] = (df_coach['cancelled'] + df_coach["refunded"]) / df_coach['total']

In [62]:
#checking
df_good_coaches=df_coach[df_coach["total"] > 100].sort_values(by="conversion rate", ascending = False)
df_good_coaches.head()

Status            cancelled  completed  refunded  total  conversion rate  \
조우찬 코치(WWN531JQ)         36         65         6    107         0.607477   
허슬지 코치(DWVG5IFL)         43         71         3    117         0.606838   
허성원 코치(9124O1IH)         43         76         7    126         0.603175   
조설영 코치(U7L98DAO)         48         78         6    132         0.590909   
권슬영 코치(E3GD4106)         42         70         9    121         0.578512   

Status            concellation rate  
조우찬 코치(WWN531JQ)           0.392523  
허슬지 코치(DWVG5IFL)           0.393162  
허성원 코치(9124O1IH)           0.396825  
조설영 코치(U7L98DAO)           0.409091  
권슬영 코치(E3GD4106)           0.421488

In [63]:
# checking
df_check_coaches =df_coach[df_coach["total"] > 100].sort_values(by="conversion rate")
df_check_coaches.head()

Status            cancelled  completed  refunded  total  conversion rate  \
조수민 코치(OBCAO3W0)         49         46         8    103         0.446602   
박도영 코치(I4KVQ5G0)         77         70         6    153         0.457516   
오동완 코치(0O48DQCH)         55         56         7    118         0.474576   
조초연 코치(3JBE9GKO)         55         56         5    116         0.482759   
오초빈 코치(A3WOLAQM)        124        150        27    301         0.498339   

Status            concellation rate  
조수민 코치(OBCAO3W0)           0.553398  
박도영 코치(I4KVQ5G0)           0.542484  
오동완 코치(0O48DQCH)           0.525424  
조초연 코치(3JBE9GKO)           0.517241  
오초빈 코치(A3WOLAQM)           0.501661